File: frontend/services/coach.ts
import { api } from "./api";

export interface ChatMessage {
  role: "user" | "assistant" | "system";
  content: string;
  timestamp: string;
}

export interface Conversation {
  _id: string;
  user_id: string;
  title: string;
  messages: ChatMessage[];
  created_at: string;
  updated_at: string;
}

export interface ChatResponse {
  response: string;
  conversation_id: string;
}

export const coachService = {
  async createConversation(title?: string): Promise<Conversation> {
    const res = await api.post<Conversation>("/coach/conversations", { title });
    return res.data;
  },

  async listConversations(search?: string): Promise<Conversation[]> {
    const params = search ? { search } : {};
    const res = await api.get<Conversation[]>("/coach/conversations", { params });
    return res.data;
  },

  async getConversation(id: string): Promise<Conversation> {
    const res = await api.get<Conversation>(`/coach/conversations/${id}`);
    return res.data;
  },

  async renameConversation(id: string, title: string): Promise<Conversation> {
    const res = await api.put<Conversation>(`/coach/conversations/${id}/rename`, { title });
    return res.data;
  },

  async deleteConversation(id: string): Promise<void> {
    await api.delete(`/coach/conversations/${id}`);
  },

  async sendMessage(id: string, message: string): Promise<ChatResponse> {
    const res = await api.post<ChatResponse>(`/coach/conversations/${id}/chat`, { message });
    return res.data;
  }
};

File: backend/schemas/coach.py
from pydantic import BaseModel, Field
from typing import List, Optional
from datetime import datetime

class MessageSchema(BaseModel):
    role: str = Field(..., description="Role of message sender (user, assistant, system)")
    content: str = Field(..., description="Content of the message")
    timestamp: datetime = Field(default_factory=datetime.utcnow)

class ConversationCreate(BaseModel):
    title: Optional[str] = Field(None, max_length=100, description="Optional title of the conversation")

class ConversationRename(BaseModel):
    title: str = Field(..., min_length=1, max_length=100, description="New title of the conversation")

class ConversationResponse(BaseModel):
    id: str = Field(..., alias="_id")
    user_id: str
    title: str
    messages: List[MessageSchema]
    created_at: datetime
    updated_at: datetime

    class Config:
        populate_by_name = True
        json_encoders = {
            datetime: lambda dt: dt.isoformat()
        }

class ChatRequest(BaseModel):
    message: str = Field(..., min_length=1, max_length=2000, description="The user's message to the AI coach")

class ChatResponse(BaseModel):
    response: str
    conversation_id: str
